toolchain/sem_ir/builtin_function_kind.cpp

```diff
@@ -0,0 +1,177 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "toolchain/sem_ir/builtin_function_kind.h"
+
+#include <utility>
+
+#include "toolchain/sem_ir/file.h"
+#include "toolchain/sem_ir/ids.h"
+#include "toolchain/sem_ir/typed_insts.h"
+
+namespace Carbon::SemIR {
+
+namespace BuiltinFunctionInfo {
+
+// A function that validates that a builtin was declared properly.
+using ValidateFn = auto(const File& sem_ir, llvm::ArrayRef<TypeId> arg_types,
+                        TypeId return_type) -> bool;
+
+namespace {
+// Information about a builtin function.
+struct BuiltinInfo {
+  llvm::StringLiteral name;
+  ValidateFn* validate;
+};
+
+// State used when validating a builtin signature that persists between
+// individual checks.
+struct ValidateState {
+  // The type `T`, if any.
+  TypeId t = TypeId::Invalid;
+};
+
+// Helper declaring a generic parameter with a constraint on its type. See
+// ValidateSignature for details.
+template <typename ParamT, typename TypeConstraint>
+struct Param {
+  static auto Check(const File& sem_ir, ValidateState& state) -> bool {
+    return TypeConstraint::Check(sem_ir, state, ParamT::Get(state));
+  }
+};
+
+// Constraint that a type is the first generic parameter of the builtin. See
+// ValidateSignature for details.
+struct T {
```

Comment:
I'm really used to "T" being a template argument. It's not obvious to me what it's doing here, or why "T" would be an appropriate name outside a template parameter. Can you elaborate?

Reply:
Maybe "AnyType" if that's appropriate?

Reply:
Here, `T` is a generic parameter of a builtin signature. I've reworked this to try to be a little clearer.